File: src/main/java/org/example/math_geometry/SpiralMatrixIII.kt
package org.example.math_geometry

fun spiralMatrixIII(rows: Int, cols: Int, rStart: Int, cStart: Int): Array<IntArray> {
    val directions = arrayOf(
        intArrayOf(0, 1),
        intArrayOf(1, 0),
        intArrayOf(0, -1),
        intArrayOf(-1, 0)
    )

    var r = rStart
    var c = cStart
    var i = 0
    var steps = 1
    val res = mutableListOf<IntArray>()
    var visited = 0
    val total = rows * cols
    while (visited < total) {
        repeat(2) {
            val (dr, dc) = directions[i]
            repeat(steps) {
                if (r in 0 until rows && c in 0 until cols) {
                    res.add(intArrayOf(r, c))
                    visited += 1
                }
                r += dr
                c += dc
            }
            i = (i + 1) % 4
        }
        steps += 1
    }

    return res.toTypedArray()
}

File: src/main/java/org/example/math_geometry/SpiralMatrixIII.py
from typing import List


class Solution:
    def spiralMatrixIII(self, rows: int, cols: int, rStart: int, cStart: int) -> List[List[int]]:
        directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]  # directions for spiral order

        i, steps = 0, 1
        res = []
        r, c = rStart, cStart
        while len(res) < rows * cols:
            # 2 because the spiral goes right and then down
            # then it goes left and up
            for _ in range(2):
                dr, dc = directions[i]
                for _ in range(steps):
                    if 0 <= r < rows and 0 <= c < cols:
                        res.append([r, c])
                    r, c = r + dr, c + dc
                i = (i + 1) % 4  # to keep it within bounds for directions
            steps += 1  # no of steps in spiral increases with each iteration
        return res


obj = Solution()
rows = 1
cols = 4
rStart = 0
cStart = 0
print(obj.spiralMatrixIII(rows, cols, rStart, cStart))